### DISCUSSION TITLE
How do you pass an entire array to an It block?

### DISCUSSION THREAD

**Discussion by codykonior** [pester/Pester#2131] [Q&A] [answered]
In the BeforeDiscovery block I'm setting up variables of lists of files to operate on. Then I have two different kinds of tests:
* The normal kind that tests each file, e.g. It -ForEach $file
* Another kind where I need the entire collection at once, e.g. checking for filenames duplicated in different folders

I can see how to do the first one in the examples:

```
BeforeDiscovery { 
  $files = ...
}

Describe "Blah" {
  It "Blah <_>" -ForEach $files {
    $file | ...
  }
}
```

But how do I do the second one? The closest I've come is to wrap it in an array again and then exclude the empty entry:

```
  It "Blah" -ForEach (, $files) {
    $files = $_ | Select-Object -Skip 1
    ... (some calculation that requires the whole array) ...
  }
```

But it's pretty kludgy. Is there another way?

**Comment by fflaten** [comment, marked as answer]
You won't get empty items with `(,$files)`, just an array with the files-array as first and only time.

```powershell
It "Blah" -ForEach (, $files) {
    # Executes once with $_ equal to the original $files
}
```

You could also use a hashtable as a wrapper. Each key will be available as a named variable in the test.

```powershell
It "Blah" -ForEach @{files = $files} {
    # executes once. $files (key) contains the array. $_ contains the hashtable 
}
```
